### PR TITLE
Optional dependent null

### PR DESCRIPTION
**Pull request by umitkavala**

### CODE REVIEW

File: test/EFCore.Relational.Specification.Tests/TableSplittingTestBase.cs [hunk 1]
```diff
@@ -548,6 +549,48 @@ public virtual async Task Optional_dependent_materialized_when_no_properties()
...
+                        ob =>
+                        {
+                            ob.ToTable("MeterReadings");
+                            ob.Property(o => o.ReadingStatus).HasColumnName("ReadingStatus");
+                            ob.HasOne(o => o.MeterReadingDetails).WithOne()
+                                .HasForeignKey<MeterReadingDetails>(o => o.Id);
+                        });
+                }
+                );
+
```
Comment (AndriySvyryd) [inline test/EFCore.Relational.Specification.Tests/TableSplittingTestBase.cs:575]: You can ignore the warning with this:

```C#
            var contextFactory = await InitializeAsync<DbContext>(
                modelBuilder => { ... },
                shouldLogCategory: _ => true,
                onConfiguring: options => options.ConfigureWarnings(w => w.Ignore(RelationalEventId.OptionalDependentWithoutIdentifyingPropertyWarning)));
```

Comment (AndriySvyryd) [reply test/EFCore.Relational.Specification.Tests/TableSplittingTestBase.cs:577]: Also add a test to make sure that the warning is not logged when there's at least one non-null value on the optional dependent.

Comment (umitkavala) [reply test/EFCore.Relational.Specification.Tests/TableSplittingTestBase.cs:577]: Working on it.

File: src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs [hunk 2]
```diff
@@ -74,6 +74,57 @@ public virtual TValue GetOrAddValue(IUpdateEntry entry, SharedTableEntryValueFac
...
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public virtual bool IsOptionalWithNull(IUpdateEntry entry)
+        {
+            var principalEntityTypesMap = new Dictionary<IEntityType, bool>();
+
+            var optional = GetPrincipalEntityTypes(entry.EntityType);
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs:87]:
```suggestion
            var optional = _table.IsOptional(entry.EntityType);
```

File: src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs [hunk 3]
```diff
@@ -74,6 +74,57 @@ public virtual TValue GetOrAddValue(IUpdateEntry entry, SharedTableEntryValueFac
...
+            var nullableWithNull = true;
+
+            if (!optional)
+            {
+                return false;
+            }
+
+            foreach (var property in entry.EntityType.GetProperties())
+            {
+                if (property.IsPrimaryKey())
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs:98]: You can skip all non-nullable properties

File: src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs (hunk 2 above)
Comment (umitkavala) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs:87]: done

File: src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs (hunk 3 above)
Comment (umitkavala) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs:98]: Also need to skip IsConcurrencyToken and entity has no nullable property at all.

Comment (umitkavala) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs:96]: This generated lots of edge cases for other tests. So I'm leaving as IsPrimarKey check. I couldn't find better way to handle other cases.

Comment (AndriySvyryd) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs:96]: You are right, to do this properly move the check to the end of `ModificationCommand.GenerateColumnModifications`, there you can throw if there were no write column modifications generated for an optional dependent.

Also if `_sensitiveLoggingEnabled` is set to `true` add the PK values to the exception message.

Comment (umitkavala) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs:96]: working on it.

Comment (umitkavala) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs]: done

Comment (AndriySvyryd) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs]: > but warning event turned into exception. is this ok?

No, this needs to be a warning, because there might be existing apps that are currently doing this without bad consequences and would ignore the warning instead of changing their logic.

Also the previous commit was pretty close to the correct conditions, just needed to tweak it slightly:

```C#
var optionalDependentWithAllNull =
    (entry.EntityState == EntityState.Deleted
        || entry.EntityState == EntityState.Added)
        && tableMapping.Table.IsOptional(entry.EntityType)
        && tableMapping.Table.GetRowInternalForeignKeys(entry.EntityType).Any();
```

```C#
if (optionalDependentWithAllNull)
{
    if (columnModification.IsWrite
        && (entry.EntityState != EntityState.Added || columnModification.Value is not null))
    {
        optionalDependentWithAllNull = false;
    }
}
```

`TPTTableSplittingSqlServerTest.Can_insert_dependent_with_just_one_parent` will fail with these changes, but you can skip it, there's an unrelated bug: https://github.com/dotnet/efcore/issues/24970

Comment (umitkavala) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs]: If we'd have a method on ModificationCommand which tells there is a warning (we'll flag it with current code instead throw)
`public bool IsOptionalWithNull()`
and we can check & throw here:
https://github.com/dotnet/efcore/blob/6449203147298f03b34d227de1a5b48f62060bf1/src/EFCore.Relational/Update/Internal/CommandBatchPreparer.cs#L93

```c#
var commandAdded = batch.AddCommand(modificationCommand)
if(modificationCommand.IsOptionalWithNull())
{
 Dependencies.UpdateLogger.OptionalDependentWithAllNullPropertiesWarning(entry.EntityType);
}
if (!commandAdded )
...

Comment (AndriySvyryd) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs]: > Changes done but I couldn't find a way to access logger to throw warning in ModificationCommand.GenerateColumnModifications.

You can add the logger to the ModificationCommand constructor and store it in a field

Comment (umitkavala) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs]: working on it.

Comment (umitkavala) [reply src/EFCore.Relational/Update/Internal/SharedTableEntryMap.cs]: done

File: src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs [hunk 4]
```diff
@@ -2964,5 +2964,61 @@ private static string OptionalDependentWithoutIdentifyingPropertyWarning(EventDe
...
+            {
+                definition.Log(diagnostics, entityType.DisplayName(), sensitiveData);
+            }
+
+            if (diagnostics.NeedsEventData(definition, out var diagnosticSourceEnabled, out var simpleLogEnabled))
+            {
+                var eventData = new EntityTypeEventData(
+                    definition,
+                    OptionalDependentWithoutIdentifyingPropertyWarning,
+                    entityType);
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs:3018]: You also need to include the sensitive data in the event data. We usually do this by passing in the entity entry and building the values string in this method.

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 5]
```diff
@@ -682,6 +686,9 @@
...
   </data>
+  <data name="OptionalDependentWithDependentWithoutIdentifyingPropertySensitive" xml:space="preserve">
+    <value>Entity type '{entityType}' with primary key values {keyValues} is an optional dependent using table sharing and containing other dependents without any required non shared property to identify whether the entity exists. If all nullable properties contain a null value in database then an object instance won't be created in the query causing nested dependent's values to be lost. Add a required property to create instances with null values for other properties or mark the incoming navigation as required to always create an instance.</value>
+  </data>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Properties/RelationalStrings.resx:691]: Delete this

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 6]
```diff
@@ -554,6 +554,10 @@
...
     <comment>Warning RelationalEventId.OptionalDependentWithoutIdentifyingPropertyWarning string</comment>
   </data>
+  <data name="LogOptionalDependentWithoutIdentifyingPropertySensitive" xml:space="preserve">
+    <value>Entity type '{entityType}' with primary key values {keyValues} is an optional dependent using table sharing and containing other dependents without any required non shared property to identify whether the entity exists. If all nullable properties contain a null value in database then an object instance won't be created in the query causing nested dependent's values to be lost. Add a required property to create instances with null values for other properties or mark the incoming navigation as required to always create an instance.</value>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Properties/RelationalStrings.resx:558]:
```suggestion
    <value>The entity of type '{entityType}' with primary key values {keyValues} is an optional dependent using table sharing. The entity does not have any property with a non-default value to identify whether the entity exists. This means that when it is queried no object instance will be created instead of an instance with all properties set to default values. Any nested dependents will also be lost. Either don't save any instance with only default values or mark the incoming navigation as required in the model.</value>
```

Also use a similar message for the non-sensitive case.

File: src/EFCore.Relational/Update/ModificationCommand.cs [hunk 7]
```diff
@@ -294,6 +299,8 @@ private IReadOnlyList<ColumnModification> GenerateColumnModifications()
                 }
             }
 
+
+
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Update/ModificationCommand.cs:303]: Remove blank lines

File: src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs (hunk 4 above)
Comment (umitkavala) [reply src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs:3018]: done

File: src/EFCore.Relational/Properties/RelationalStrings.resx (hunk 5 above)
Comment (umitkavala) [reply src/EFCore.Relational/Properties/RelationalStrings.resx:691]: done

File: src/EFCore.Relational/Properties/RelationalStrings.resx (hunk 6 above)
Comment (umitkavala) [reply src/EFCore.Relational/Properties/RelationalStrings.resx:558]: done

File: src/EFCore.Relational/Update/ModificationCommand.cs (hunk 7 above)
Comment (umitkavala) [reply src/EFCore.Relational/Update/ModificationCommand.cs:303]: done

File: src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs [hunk 8]
```diff
@@ -2964,5 +2964,69 @@ private static string OptionalDependentWithoutIdentifyingPropertyWarning(EventDe
                 diagnostics.DispatchEventData(definition, eventData, diagnosticSourceEnabled, simpleLogEnabled);
             }
         }
+        
+        /// <summary>
+        ///     Logs the <see cref="RelationalEventId.OptionalDependentWithoutIdentifyingPropertyWarning" /> event.
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs:2969]: Add a new property to `RelationalEventId` for these events

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 9]
```diff
@@ -551,7 +551,11 @@
     <comment>Debug RelationalEventId.ConnectionOpening string string</comment>
   </data>
   <data name="LogOptionalDependentWithoutIdentifyingProperty" xml:space="preserve">
-    <value>The entity type '{entityType}' is an optional dependent using table sharing without any required non shared property that could be used to identify whether the entity exists. If all nullable properties contain a null value in database then an object instance won't be created in the query. Add a required property to create instances with null values for other properties or mark the incoming navigation as required to always create an instance.</value>
+    <value>The entity of type '{entityType}' is an optional dependent using table sharing. The entity does not have any property with a non-default value to identify whether the entity exists. This means that when it is queried no object instance will be created instead of an instance with all properties set to default values. Any nested dependents will also be lost. Either don't save any instance with only default values or mark the incoming navigation as required in the model.</value>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Properties/RelationalStrings.resx:554]: Don't modify the existing event for `OptionalDependentWithoutIdentifyingProperty`, add a new entry for `OptionalDependentWithAllNullPropertiesWarning`

File: src/EFCore/Diagnostics/EntityTypeEventData.cs [hunk 10]
```diff
@@ -28,9 +29,29 @@ public class EntityTypeEventData : EventData
...
+        /// <summary>
+        ///     Constructs the event payload.
+        /// </summary>
+        /// <param name="eventDefinition"> The event definition. </param>
+        /// <param name="messageGenerator"> A delegate that generates a log message for this event. </param>
+        /// <param name="entityEntry"> The entry for the entity instance on which the property value has changed. </param>
+        public EntityTypeEventData(
+            EventDefinitionBase eventDefinition,
+            Func<EventDefinitionBase, EventData, string> messageGenerator,
+            IUpdateEntry? entityEntry)
```
Comment (AndriySvyryd) [inline src/EFCore/Diagnostics/EntityTypeEventData.cs:41]: Create `UpdateEntryEventData` instead of adding this to `EntityTypeEventData`

File: test/EFCore.Specification.Tests/TestModels/QueryModel/SharedTableContext.cs [hunk 11]
```diff
@@ -0,0 +1,19 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.EntityFrameworkCore.TestUtilities;
+
+
+namespace Microsoft.EntityFrameworkCore.TestModels.QueryModel
+{
+    public class SharedTableContext : PoolableDbContext
```
Comment (AndriySvyryd) [inline test/EFCore.Specification.Tests/TestModels/QueryModel/SharedTableContext.cs:9]: It's unlikely that this model would be used in other tests. Make them protected nested classes in `TableSplittingTestBase` or you can delete them and rewrite the tests using `TransportationContext`

File: src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs (hunk 8 above)
Comment (umitkavala) [reply src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs:2969]: done

File: src/EFCore.Relational/Properties/RelationalStrings.resx (hunk 9 above)
Comment (umitkavala) [reply src/EFCore.Relational/Properties/RelationalStrings.resx:554]: done

File: test/EFCore.Specification.Tests/TestModels/QueryModel/SharedTableContext.cs (hunk 11 above)
Comment (umitkavala) [reply test/EFCore.Specification.Tests/TestModels/QueryModel/SharedTableContext.cs:9]: move as protected class. I couldn't find a way to use transportation models as optional dependent.

File: src/EFCore/Diagnostics/EntityTypeEventData.cs (hunk 10 above)
Comment (umitkavala) [reply src/EFCore/Diagnostics/EntityTypeEventData.cs]: done

File: src/EFCore.Relational/Diagnostics/RelationalEventId.cs [hunk 12]
```diff
@@ -738,6 +739,20 @@ private static EventId MakeValidationId(Id id)
         public static readonly EventId OptionalDependentWithoutIdentifyingPropertyWarning
             = MakeValidationId(Id.OptionalDependentWithoutIdentifyingPropertyWarning);
 
+        /// <summary>
+        ///     <para>
+        ///         A foreign key specifies properties which don't map to the related tables.
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/RelationalEventId.cs:744]:
```suggestion
        ///         The entity does not have any property with a non-default value to identify whether the entity exists.
```

File: src/EFCore/Diagnostics/UpdateEntryEventData.cs [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+using System;
+using System.Diagnostics;
+using Microsoft.EntityFrameworkCore.Update;
+
+namespace Microsoft.EntityFrameworkCore.Diagnostics
+{
+    /// <summary>
+    ///     A <see cref="DiagnosticSource" /> event payload class for events that have
+    ///     an entity type.
```
Comment (AndriySvyryd) [inline src/EFCore/Diagnostics/UpdateEntryEventData.cs:12]:
```suggestion
    ///     an entity update entry.
```

File: src/EFCore.Relational/Diagnostics/RelationalEventId.cs [hunk 14]
```diff
@@ -738,6 +739,20 @@ private static EventId MakeValidationId(Id id)
...
+        ///     </para>
+        ///     <para>
+        ///         This event is in the <see cref="DbLoggerCategory.Update" /> category.
+        ///     </para>
+        ///     <para>
+        ///         This event uses the <see cref="EntityTypeEventData" /> payload when used with a <see cref="DiagnosticSource" />.
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/RelationalEventId.cs:750]:
```suggestion
        ///         This event uses the <see cref="UpdateEntryEventData" /> payload when used with a <see cref="DiagnosticSource" />.
```

File: src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs [hunk 15]
```diff
@@ -2964,5 +2964,101 @@ private static string OptionalDependentWithoutIdentifyingPropertyWarning(EventDe
                 diagnostics.DispatchEventData(definition, eventData, diagnosticSourceEnabled, simpleLogEnabled);
             }
         }
+
+        /// <summary>
+        ///     Logs the <see cref="RelationalEventId.OptionalDependentWithoutIdentifyingPropertyWarning" /> event.
+        /// </summary>
+        /// <param name="diagnostics"> The diagnostics logger to use. </param>
+        /// <param name="entityType"> The entity type. </param>
+        public static void OptionalDependentWithoutIdentifyingPropertyWarning(
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs:2973]: Remove this overload

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 16]
```diff
@@ -550,6 +550,14 @@
     <value>Opening connection to database '{database}' on server '{server}'.</value>
     <comment>Debug RelationalEventId.ConnectionOpening string string</comment>
   </data>
+  <data name="LogOptionalDependentWithAllNullProperties" xml:space="preserve">
+    <value>The entity of type '{entityType}' is an optional dependent using table sharing. The entity does not have any property with a non-default value to identify whether the entity exists. This means that when it is queried no object instance will be created instead of an instance with all properties set to default values. Any nested dependents will also be lost. Either don't save any instance with only default values or mark the incoming navigation as required in the model.</value>
+    <comment>Warning RelationalEventId.OptionalDependentWithoutIdentifyingPropertyWarning string</comment>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Properties/RelationalStrings.resx:555]:
```suggestion
    <comment>Warning RelationalEventId.OptionalDependentWithAllNullPropertiesWarning string</comment>
```

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 17]
```diff
@@ -550,6 +550,14 @@
...
+  </data>
+  <data name="LogOptionalDependentWithAllNullPropertiesSensitive" xml:space="preserve">
+    <value>The entity of type '{entityType}' with primary key values {keyValues} is an optional dependent using table sharing. The entity does not have any property with a non-default value to identify whether the entity exists. This means that when it is queried no object instance will be created instead of an instance with all properties set to default values. Any nested dependents will also be lost. Either don't save any instance with only default values or mark the incoming navigation as required in the model.</value>
+    <comment>Warning RelationalEventId.OptionalDependentWithoutIdentifyingPropertyWarning string</comment>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Properties/RelationalStrings.resx:559]:
```suggestion
    <comment>Warning RelationalEventId.OptionalDependentWithAllNullPropertiesWarning string</comment>
```

File: src/EFCore.Relational/Properties/RelationalStrings.Designer.cs [hunk 18]
```diff
@@ -2214,7 +2222,7 @@ public static EventDefinition<string> LogNoMigrationsFound(IDiagnosticsLogger lo
         }
 
         /// <summary>
-        ///     The entity type '{entityType}' is an optional dependent using table sharing without any required non shared property that could be used to identify whether the entity exists. If all nullable properties contain a null value in database then an object instance won't be created in the query. Add a required property to create instances with null values for other properties or mark the incoming navigation as required to always create an instance.
+        ///     The entity of type '{entityType}' is an optional dependent using table sharing. The entity does not have any property with a non-default value to identify whether the entity exists. This means that when it is queried no object instance will be created instead of an instance with all properties set to default values. Any nested dependents will also be lost. Either don't save any instance with only default values or mark the incoming navigation as required in the model.
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Properties/RelationalStrings.Designer.cs:2225]: Use "Run custom tool" to regenerate this file

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 19]
```diff
@@ -550,6 +550,14 @@
     <value>Opening connection to database '{database}' on server '{server}'.</value>
     <comment>Debug RelationalEventId.ConnectionOpening string string</comment>
   </data>
+  <data name="LogOptionalDependentWithAllNullProperties" xml:space="preserve">
+    <value>The entity of type '{entityType}' is an optional dependent using table sharing. The entity does not have any property with a non-default value to identify whether the entity exists. This means that when it is queried no object instance will be created instead of an instance with all properties set to default values. Any nested dependents will also be lost. Either don't save any instance with only default values or mark the incoming navigation as required in the model.</value>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Properties/RelationalStrings.resx:554]:
```suggestion
    <value>The entity of type '{entityType}' is an optional dependent using table sharing. The entity does not have any property with a non-default value to identify whether the entity exists. This means that when it is queried no object instance will be created instead of an instance with all properties set to default values. Any nested dependents will also be lost. Either don't save any instance with only default values or mark the incoming navigation as required in the model. Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the key values of the entity.</value>
```

File: src/EFCore.Relational/Properties/RelationalStrings.Designer.cs (hunk 18 above)
Comment (umitkavala) [reply src/EFCore.Relational/Properties/RelationalStrings.Designer.cs:2225]: done

File: src/EFCore.Relational/Properties/RelationalStrings.resx (hunk 19 above)
Comment (umitkavala) [reply src/EFCore.Relational/Properties/RelationalStrings.resx:554]: done

File: src/EFCore.Relational/Properties/RelationalStrings.resx (hunk 17 above)
Comment (umitkavala) [reply src/EFCore.Relational/Properties/RelationalStrings.resx:559]: done

File: src/EFCore.Relational/Properties/RelationalStrings.resx (hunk 16 above)
Comment (umitkavala) [reply src/EFCore.Relational/Properties/RelationalStrings.resx:555]: done

File: src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs (hunk 15 above)
Comment (umitkavala) [reply src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs:2973]: done

File: src/EFCore.Relational/Diagnostics/RelationalEventId.cs (hunk 14 above)
Comment (umitkavala) [reply src/EFCore.Relational/Diagnostics/RelationalEventId.cs:750]: done

File: src/EFCore/Diagnostics/UpdateEntryEventData.cs (hunk 13 above)
Comment (umitkavala) [reply src/EFCore/Diagnostics/UpdateEntryEventData.cs:12]: done

File: src/EFCore.Relational/Diagnostics/RelationalEventId.cs (hunk 12 above)
Comment (umitkavala) [reply src/EFCore.Relational/Diagnostics/RelationalEventId.cs:744]: done

File: src/EFCore.Relational/Diagnostics/RelationalEventId.cs [hunk 20]
```diff
@@ -738,6 +739,20 @@ private static EventId MakeValidationId(Id id)
...
+        ///     </para>
+        ///     <para>
+        ///         This event is in the <see cref="DbLoggerCategory.Update" /> category.
+        ///     </para>
+        ///     <para>
+        ///         This event uses the <see cref="UpdateEntryEventData" /> payload when used with a <see cref="DiagnosticSource" />.
+        ///     </para>
+        /// </summary>
+        public static readonly EventId OptionalDependentWithAllNullPropertiesWarning
+            = MakeValidationId(Id.OptionalDependentWithAllNullPropertiesWarning);
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/RelationalEventId.cs:754]:
```suggestion
            = MakeUpdateId(Id.OptionalDependentWithAllNullPropertiesWarning);
```

File: src/EFCore.Relational/Diagnostics/RelationalEventId.cs [hunk 21]
```diff
@@ -88,6 +88,7 @@ private enum Id
             IndexPropertiesMappedToNonOverlappingTables,
             ForeignKeyPropertiesMappedToUnrelatedTables,
             OptionalDependentWithoutIdentifyingPropertyWarning,
+            OptionalDependentWithAllNullPropertiesWarning,
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/RelationalEventId.cs:91]: Move this to line 98

File: src/EFCore.Relational/Diagnostics/RelationalEventId.cs (hunk 20 above)
Comment (umitkavala) [reply src/EFCore.Relational/Diagnostics/RelationalEventId.cs:754]: done

File: src/EFCore.Relational/Diagnostics/RelationalEventId.cs (hunk 21 above)
Comment (umitkavala) [reply src/EFCore.Relational/Diagnostics/RelationalEventId.cs:91]: done

File: src/EFCore.Relational/Diagnostics/RelationalEventId.cs [hunk 22]
```diff
@@ -738,6 +739,20 @@ private static EventId MakeValidationId(Id id)
...
+        ///         The entity does not have any property with a non-default value to identify whether the entity exists.
+        ///     </para>
+        ///     <para>
+        ///         This event is in the <see cref="DbLoggerCategory.Update" /> category.
+        ///     </para>
+        ///     <para>
+        ///         This event uses the <see cref="UpdateEntryEventData" /> payload when used with a <see cref="DiagnosticSource" />.
+        ///     </para>
+        /// </summary>
+        public static readonly EventId OptionalDependentWithAllNullPropertiesWarning
```
[the remark this answers is not in the document]
Comment (umitkavala) [reply src/EFCore.Relational/Diagnostics/RelationalEventId.cs]: done

**Review by AndriySvyryd — APPROVED** [review]
Thanks for your contribution!
